### PR TITLE
*: Reorg part update table

### LINKED ISSUE #38535: Support REORGANIZE PARTITION Key Dev Task
# Support Reorganize Partition Dev Task

Related Issues: #15000
Design PR:
- [x] https://github.com/pingcap/tidb/pull/38246

## Dev Tasks
- [x] Parser support
- [x] Copy records, #38460
- [x] Recreate the indexes in the new partitions, #38460
- [x] Double write data in foreground sessions (concurrent DML during DDL), #38508
- [x] Handle `reorganize partition` command in ddl_api.go and ddl.go, #40473
- [x] Error handling and DDL rollback, #40543
- [x] Delete range of old partitions, #40543
- [x] Update Table handling, #40714
- [ ] TiFlash handling, #40715
- [x] Add telemetry for reorganize partition, #40258
- [ ] Support in TiFlash https://github.com/pingcap/tiflash/issues/6444
- [ ] Support in TiCDC https://github.com/pingcap/tiflow/pull/8114

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

PR on top of https://github.com/pingcap/tidb/pull/40543 so looks bigger than it is.

Implementing Reorganize Partition specifics of Update Table (+ updated test)

Issue Number: close #38535

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: statistics/handle/ddl.go [hunk 1]
```diff
@@ -61,6 +61,21 @@ func (h *Handle) HandleDDLEvent(t *util.Event) error {
 				return err
 			}
 		}
+	case model.ActionReorganizePartition:
+		for _, def := range t.PartInfo.Definitions {
+			// TODO: Should we trigger analyze instead of adding 0s?
```
Comment (bb7133) [inline statistics/handle/ddl.go:66]: What does 'adding 0s' mean?

File: ddl/reorg.go [hunk 2]
```diff
@@ -553,12 +569,12 @@ func getTableRange(ctx *JobContext, d *ddlCtx, tbl table.PhysicalTable, snapshot
...
-		logutil.BgLogger().Info("[ddl] get noop table range",
+		logutil.BgLogger().Info("[ddl] get table range, endHandle < startHandle",
 			zap.String("table", fmt.Sprintf("%v", tbl.Meta())),
 			zap.Int64("table/partition ID", tbl.GetPhysicalID()),
-			zap.String("start key", hex.EncodeToString(startHandleKey)),
-			zap.String("end key", hex.EncodeToString(endHandleKey)),
-			zap.Bool("is empty table", isEmptyTable))
+			zap.Bool("isEmptyTable", isEmptyTable),
+			zap.String("endHandle", hex.EncodeToString(endHandleKey)),
+			zap.String("startHandle", hex.EncodeToString(startHandleKey)))
```
Comment (tangenta) [inline ddl/reorg.go:577]: This seems to be an unexpected change?

File: ddl/backfilling.go [hunk 3]
```diff
@@ -610,7 +610,6 @@ func (dc *ddlCtx) sendTasksAndWait(scheduler *backfillScheduler, totalAddedCount
 		return errors.Trace(err)
 	}
 
-	// nextHandle will be updated periodically in runReorgJob, so no need to update it here.
```
[the remark this answers is not in the document]
Comment (mjonss) [reply ddl/backfilling.go:613]: Right, it should have been removed in https://github.com/pingcap/tidb/pull/38738, when it was changed.

File: ddl/reorg.go (hunk 2 above)
Comment (mjonss) [reply ddl/reorg.go:577]: Right, it was changed during debugging, let me remove that change.

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**